lib/state-store.js

```diff
@@ -859,26 +859,40 @@ class StateStore {
    * @param {string} checkpoint - The sequence number to store as the recovery point.
    * @param {string} shardsPath - The path pointing to where the shards state is stored.
    * @param {Object} shardsPathNames - The values of the attribute names in the path.
+   * @param {Date} timestamp - The approximate arrival timestamp
    * @fulfil {undefined}
    * @returns {Promise}
    */
-  async storeShardCheckpoint(shardId, checkpoint, shardsPath, shardsPathNames) {
+  async storeShardCheckpoint(shardId, checkpoint, shardsPath, shardsPathNames, timestamp) {
+    const privateProps = internal(this);
+    const { logger } = privateProps;
     if (typeof checkpoint !== 'string') throw new TypeError('The sequence number is required.');
+    let timestampString = null;
+    if (timestamp instanceof Date) {
+      timestampString = timestamp.toISOString();
+    } else if (timestamp) {      
+      logger.warn("Will ignore timestamp provided, it was not a date: " + JSON.stringify(timestamp)); 
+    } else {
+      logger.warn("Will ignore timestamp provided, it was null"); 
+    }
+
     const { client, consumerGroup, streamName } = internal(this);
 
     await client.update({
       ExpressionAttributeNames: {
         ...shardsPathNames,
         '#b': shardId,
         '#c': 'checkpoint',
-        '#d': 'version'
+        '#d': 'version',
+        '#e': 'timestamp'
```

Comment:
Please make the version attribute be the last one (minor optimization). And I wouldn't consider calling it just `timestamp` but `approxArrival` instead so it's more understandable for folks that know Kinesis.

Reply:
`lastApproxArrival` etc. not just `timestamp`